lib/jekyll/filters.rb

```diff
@@ -389,7 +389,11 @@ def groupable?(element)
     private
     def item_property(item, property)
       if item.respond_to?(:to_liquid)
-        item.to_liquid[property.to_s]
+        res = item.to_liquid
+        property.to_s.split(".").each do |attr|
+          res = res[attr]
+        end
+        res
       elsif item.respond_to?(:data)
         item.data[property.to_s]
```

Comment:
@ashmaroli You are right, it wasn't working for data files.

Here are my tests:
`_data/test/test.json`
```
{
	"array": [
		{"header": { "year": 2001 }},
		{"header": { "year": 2004 }},
		{"header": { "year": 2006 }},
		{"header": { "year": 1972 }},
		{"header": { "year": 2010 }},
		{"header": { "year": 1990 }},
		{"header": { "year": 12 }},
		{"header": { "year": 0 }},
		{"header": { "year": 1996 }}
	]
}
```
`index.html`
```
---
---
{% assign test = site.data.test.array | sort: 'header.year' %}
{{ test }}
{% for i in test %}
	{{ i.header.year  }}
{% endfor %}
```

Commands : 
```
$ jekyll -v
jekyll 3.4.3
```
```
$ jekyll build && cat _site/index.html
Configuration file: none
            Source: /tmp/test-site
       Destination: /tmp/test-site/_site
 Incremental build: disabled. Enable with --incremental
      Generating...
                    done in 0.01 seconds.
 Auto-regeneration: disabled. Use --watch to enable.

{"header"=>{"year"=>2001}}{"header"=>{"year"=>2004}}{"header"=>{"year"=>2006}}{"header"=>{"year"=>1972}}{"header"=>{"year"=>2010}}{"header"=>{"year"=>1990}}{"header"=>{"year"=>12}}{"header"=>{"year"=>0}}{"header"=>{"year"=>1996}}

        2001

        2004

        2006

        1972

        2010

        1990

        12

        0

        1996

```

With the last commit :
```
$ ../jekyll/exe/jekyll build && cat _site/index.html
Configuration file: none
            Source: /tmp/test-site
       Destination: /tmp/test-site/_site
 Incremental build: disabled. Enable with --incremental
      Generating...
                    done in 0.009 seconds.
 Auto-regeneration: disabled. Use --watch to enable.

{"header"=>{"year"=>0}}{"header"=>{"year"=>12}}{"header"=>{"year"=>1972}}{"header"=>{"year"=>1990}}{"header"=>{"year"=>1996}}{"header"=>{"year"=>2001}}{"header"=>{"year"=>2004}}{"header"=>{"year"=>2006}}{"header"=>{"year"=>2010}}

        0

        12

        1972

        1990

        1996

        2001

        2004

        2006

        2010

```